types/styled-components/index.d.ts

```diff
@@ -59,32 +60,47 @@ export type ThemedGlobalStyledClassProps<P, T> = P & {
 export interface GlobalStyleClass<P, T>
     extends React.ComponentClass<ThemedGlobalStyledClassProps<P, T>> {}
 
-export interface StyledComponentClass<P, T, O = P>
-    extends React.ComponentClass<ThemedOuterStyledProps<O, T>> {
+/** @deprecated Use StyledComponent instead */
+export type StyledComponentClass<P, T, O = {}> = StyledComponent<P, T, O>;
+
+// Make the StyledComponentInterpolation non-callable because typescript can't tell
+// between this and InterpolationFunction when inferring Interpolation
+type StyledComponentInterpolation = Pick<StyledComponent<any, any>, keyof StyledComponent<any, any>>;
```

Comment:
Since with the ExoticComponent there were 2 function signatures in `InterpolationValue` the compiler just couldn't pick any and turned all of the interpolated functions into (fortunately signalling) implicit anys.